interpreter/cling/lib/Interpreter/LookupHelper.cpp

```diff
@@ -1211,6 +1205,9 @@ namespace cling {
             S.InstantiateFunctionDefinition(SourceLocation(), TheDecl,
                                             true /*recursive instantiation*/);
           }
+          if (S.getDiagnostics().hasErrorOccurred()) {
+            TheDecl->setInvalidDecl();
```

Comment:
That won't be enough: any instantiation of templates that this template instantiation depends on might have failed, and an unknown subset of the instantiations might need to get unloaded... We might need to listen to the AST operations during lookup and roll everything back. (Not as part of this PR.) @jalopezg-git thoughts?

Reply:
Maybe starting a (nested) transaction and unloading the decl group represented by it?

Reply:
There is already a transaction which gets started in this case.  I've modified the logic to catch the error and fail the transaction rather than unloading the decl directly.